polar-core/src/kb.rs

```diff
@@ -1,26 +1,42 @@
 use std::collections::HashMap;
 
-use crate::error::PolarResult;
+use crate::error::ParameterError;
+use crate::error::{PolarError, PolarResult};
 
 pub use super::bindings::Bindings;
 use super::counter::Counter;
 use super::rules::*;
 use super::sources::*;
 use super::terms::*;
+use std::sync::Arc;
 
-/// A map of bindings: variable name → value. The VM uses a stack internally,
-/// but can translate to and from this type.
+enum RuleParamMatch {
+    True,
+    False(String),
+}
+
+impl RuleParamMatch {
+    #[cfg(test)]
+    fn is_true(&self) -> bool {
+        matches!(self, RuleParamMatch::True)
+    }
+}
 
 #[derive(Default)]
 pub struct KnowledgeBase {
+    /// A map of bindings: variable name → value. The VM uses a stack internally,
+    /// but can translate to and from this type.
     pub constants: Bindings,
+    /// Map of class name -> MRO list where the MRO list is a list of class instance IDs
+    mro: HashMap<Symbol, Vec<u64>>,
 
     /// Map from loaded files to the source ID
     pub loaded_files: HashMap<String, u64>,
     /// Map from source code loaded to the filename it was loaded as
     pub loaded_content: HashMap<String, String>,
 
-    pub rules: HashMap<Symbol, GenericRule>,
+    rules: HashMap<Symbol, GenericRule>,
+    rule_prototypes: HashMap<Symbol, GenericRule>,
```

Comment:
Left a comment about this before, but I can't find the resolution. Why can't this just be `HashMap<Symbol, Vec<Rule>>`?

Reply:
There's not a real reason. Can change it if you want.

Reply:
The rationale for changing is generic rule has a rule index for efficiently looking up rules in the VM. We don't use that index here so we aren't really getting any benefit from it.

You can decide if you want to change it or not.